### ISSUE TITLE
[Bug] Very slow, choppy performance of SVG compared to Chrome on same device (example included)

### ISSUE THREAD

**Issue by psy0rz** [mozilla-mobile/fenix#17216]
(This is an existing bug that unfortunately still exists: https://bugzilla.mozilla.org/show_bug.cgi?id=1289866 )

Visit this page on both Firefox nightly (fenix) and Chrome on Android: http://vallandingham.me/gates_bubbles/

Performance on Fenix is much slower.

I'm designing an online collaborate drawing tool, and Fenix is completely unusable compare to a smooth experience in Chrome. I know my application is pushing the limits sort of, so thats why I included the more simple example above. Alpha demo of our app: http://internet-paper.com/paper.html  (sources https://github.com/psy0rz/etherpaint )

Try pinch-zooming on both Chrome and Fenix. Chrome has 10+fps vs Fenix  <1fps.

The performance difference between the two are orders of magnitude. 

Optimization tips or  workarounds are welcome as well. (we're going to use simplyjs to reduce the polylines) 

* Android device: LG Velvet
* Fenix version: Nightly 201224 17:00

**Comment by psy0rz** [comment]
I also tried remote performance analyzing the regular Firefox via remote USB: It shows 80% idle while it also shows almost 0fps. So everything seems fast enough, yet for some reason its not.

**Comment by Dunexus** [comment]
Hello,
Do you have webrender enabled ?

I tried on Oneplus 5T (Android 10) with webrender enabled and it is quite smooth.

**Comment by sheikh-azharuddin** [comment]
Seems malicious ads and webpages embedded in the link

Careful guys

![Screenshot_20201225-214630860](https://user-images.githubusercontent.com/58127841/103138657-b5c09e80-46fa-11eb-929f-87dc061b3c46.jpg)

**Comment by MicroBeast** [comment]
That's not coming from any links from OP.

**Comment by sheikh-azharuddin** [comment]
> That's not coming from any links from OP.

Visit the link http://internet-paper.com/paper.html
And click on any grid whitespace in middle of screen... Ads will be bombarded

**Comment by psy0rz** [comment]
> Seems malicious ads and webpages embedded in the link
> 
> Careful guys
> 
> ![Screenshot_20201225-214630860](https://user-images.githubusercontent.com/58127841/103138657-b5c09e80-46fa-11eb-929f-87dc061b3c46.jpg)

I assure you, that shouldnt come from my html code. Or it should be from one of the external assets which i highly doubt. (in the current masterbranch the assets are hosted locally from now on)

However i see that the server has crashed as well, so i'll investigate that.

**Comment by psy0rz** [comment]
> Hello,
> Do you have webrender enabled ?
> 
> I tried on Oneplus 5T (Android 10) with webrender enabled and it is quite smooth.

I first had to find out how to able it and how to get to about:config (available in Firefox beta only for some reason).

Then i checked about:support on Fenix to discover that compositing already was WebRender on my Fenix by default.

But still zooming on http://internet-paper.com/paper.html is super slow. (and even if it wasnt: telling our users to enable some option via about:config is not a long term solution offcourse)

For now we have to tell our users that Firefox mobile isn't supported and they have to use something else. Fenix still seems a long way out and i'm afraid this issue will never get fixed anyway. The original bug report at https://bugzilla.mozilla.org/show_bug.cgi?id=1289866 was 5 years ago..

**Comment by sheikh-azharuddin** [comment]
Page redirecting to suspicious sites...
Just zoom in the page...click on empty grid on right hand side..it is showing. @MicroBeast 

![2020-12-26-14-32-16](https://user-images.githubusercontent.com/58127841/103148593-8310b780-4787-11eb-9ac0-47a692d8a259.gif)

**Comment by sheikh-azharuddin** [comment]
> Hello,
> Do you have webrender enabled ?
> 
> I tried on Oneplus 5T (Android 10) with webrender enabled and it is quite smooth.

How to enable webrender...I mean I enabled GFX.webrender.all to true but still zooming in and out too slow..it is faster in chrome

**Comment by psy0rz** [comment]
Thanks for confirming the slowness. 

I still don't understand whats happening with that redirect. I think its your phone or network perhaps. Also check my github profile to assure yourself that i have nothing to gain from redirecting to some lame spam site.

Keep in mind the connection to that site is encrypted yet, so something along the way could be inserting ads as well.

**Comment by psy0rz** [comment]
Since its XMAS i fiddled some more to pinpoint the problem and allow better test comparisons :)

http://internet-paper.com/svgtest.html

This actually crashes firefox on the desktop and mobile, but runs pretty ok on Chromium.

What it does is: 

- Draw 200 random polylines with 200 segments each with a max size of 10000x10000. (no problems here)
- Randomly zoom and pan, inside an animation frame. 
- Display the time difference between the animation frames. (16ms on a 60hz monitor = perfect)

It seems viewbox zoom is much faster than element resizing, so I will use that from now on. It also seems nesting svg's makes things slower, so i'll try and see if I can remove that as well. 

Since i need to optimize for performance anyway, i might post some more info later.

I hope this helps to track down the bottleneck.

Edwin

**Comment by sheikh-azharuddin** [comment]
> Thanks for confirming the slowness.
> 
> I still don't understand whats happening with that redirect. I think its your phone or network perhaps. Also check my github profile to assure yourself that i have nothing to gain from redirecting to some lame spam site.
> 
> Keep in mind the connection to that site is not encrypted yet, so something along the way could be inserting ads as well.

Might be bro...first time I got this error with this page..seems someone injected something

**Comment by psy0rz** [comment]
update: it doesnt crash, it just eats 10Gb of memory and then hits the OOM killer.

**Comment by psy0rz** [comment]
I changed the zoom to be as simple a possible, only an SVG element and viewbox zooming: http://internet-paper.com/svgtest2.html

The firefox variants only seem to be slower with this specific set of circumstances. They are orders of magnitude slower when the polygons have very long segments. 

All and all much faster and hopefully acceptable for our use case. 

TLDR
-  huge polygons have slow viewbox zooming compared to chrome. 
- zooming by resizing the element instead of a viewbox zoom can be catastrophic. (while still smooth on chrome)
- tested with firefox android/firefox beta android/fenix/firefox desktop on ubuntu 20.

**Comment by mcomella** [comment]
Thanks for reporting – our platform performance team works on Bugzilla so we'll refile the issue there to get someone to look into it. Sorry for the delay in responding to the issue (due to the holiday).

**Comment by mcomella** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1686590

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).

**Comment by psy0rz** [comment]
Wait what??

The original bug WAS on mozilla and moved from https://bugzilla.mozilla.org/show_bug.cgi?id=1289866 TO github:

> We have completed our launch of our new Firefox on Android. The development of the new versions use GitHub for issue tracking. If the bug report still reproduces in a current version of [Firefox on Android nightly](https://play.google.com/store/apps/details?id=org.mozilla.fenix) an issue can be reported at the [Fenix GitHub project](https://github.com/mozilla-mobile/fenix/). If you want to discuss your report please use [Mozilla's chat](https://wiki.mozilla.org/Matrix#Connect_to_Matrix) server https://chat.mozilla.org and join the [#fenix](https://chat.mozilla.org/#/room/#fenix:mozilla.org) channel.

Now you've moved it to mozilla again? 

No wonder firefox or whatever its called is such a mess..
